### PR TITLE
store,imagestore: treat `graphRoot` as `rwImageStore` to support `Delete`

### PR DESCRIPTION
**Pull request by flouthoc**
PR https://github.com/containers/storage/pull/1549 added support for split-filesystem using `--imagestore` and when `--imagestore` is used current `graphRoot` is mounted as an additional store.

In such case the `graphRoot` must be mounted as an `rw` additional image store so users can still `Delete` images from the `graphRoot` if they want to .

### CODE REVIEW

File: tests/split-store.bats [hunk 1]
```diff
@@ -89,9 +89,6 @@ load helpers
 	[[ "$output" =~ "Name: $name" ]]
 	[[ "$output" =~ "Data: random1" ]]
 	[[ "$output" =~ "Data: random2" ]]
-	# Since this image is being read from the readonly graph root
-	# so it must show that
-	[[ "$output" =~ "Read Only: true" ]]
```
Comment (giuseppe) [inline tests/split-store.bats:94]: Could we test it is writeable instead?

**Review by giuseppe — APPROVED** [review]
LGTM just one nit

File: tests/split-store.bats (hunk 1 above)
Comment (flouthoc) [reply tests/split-store.bats:94]: Modified the test to verify if we can actually delete image from `graphRoot` while `--image-store` is still set.

**Review by giuseppe — APPROVED** [review]
LGTM

**Comment by flouthoc** [comment]
@rhatdan @nalind PTAL

File: store.go [hunk 2]
```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
...
+		// write-able image store.
+		imageStores := []rwImageStore{s.imageStore}
+		imageStores = append(imageStores, s.rwImageStores...)
+		imageFound := false
+		primaryImageStore := true
+		for _, imageStore := range imageStores {
+			if !imageStore.Exists(id) {
+				primaryImageStore = false
+				continue
+			}
```
Comment (TomSweeneyRedHat) [inline store.go:2507]: If we go through this for loop 3 times, and the 2nd iteration we hit the continue , but the third we find it, imageFound will be set to true.  Is that valid in that circumstance?  (Stretching my store memory greatly).

Comment (flouthoc) [reply store.go:2507]: Yes that sounds valid.

**Comment by rhatdan** [comment]
LGTM

File: store.go [hunk 3]
```diff
@@ -940,9 +941,22 @@ func (s *store) load() error {
...
 	for _, store := range driver.AdditionalImageStores() {
 		gipath := filepath.Join(store, driverPrefix+"images")
-		ris, err := newROImageStore(gipath)
-		if err != nil {
-			return err
+		var ris roImageStore
+		if s.imageStoreDir != "" && store == s.graphRoot {
+			// If --imagestore was set and current store
+			// is `graphRoot` then mount it as a `rw` additional
+			// store instead of `readonly` additional store.
```
Comment (mtrmac) [inline store.go:948]: The code says this already.

It would be much more useful to have a comment saying _why_ this is necessary.

File: store.go [hunk 4]
```diff
@@ -676,6 +676,7 @@ type store struct {
 	autoNsMinSize   uint32
 	autoNsMaxSize   uint32
 	imageStore      rwImageStore
+	rwImageStores   []rwImageStore
```
Comment (mtrmac) [inline store.go:679]: If we need to special-case this situation, fine.

This `rwImageStores`:
- has only zero or one element
- does not directly correspond to any user-visible UI feature

So to me, that’s an a surprising way to record the data. It seems to me to be an unnecessary generalization just complicating the code right now.

_At the very least_ this field needs documentation as to how it relates to `imageStore`/`roImageStores` (it _looks_ as if it is some kind of direct sibling to `imageStore`/`roImageStore`, which is not the case). And I’d prefer the relevant stores to be somehow directly modeled here — “primary image store” / “graph-root image store” / “additional image stores as configured by user”, or something like that.

---

To my way of thinking, it’s surprising, inconsistent and hard to follow that the responsibility for `options.ImageStore` is so diffuse. This `store` object is responsible for creating the directories based on `options.ImageStore`, but somehow the individual graph driver is responsible for turning `options.ImageStore` into an entry in `AdditionalImageStores`, but it _must_ use precisely the `s.graphRoot` path, or the upper-level `store` breaks.

This is not at all a clean division of responsibility between the top-level store and the graph driver. I feel rather confident in predicting that it will lead to bugs.

(To be fair, all of the “additional * store” has similar design entanglement between `store` and the overlay driver; but this is expanding it to a much larger degree.)

---

Either this whole image store feature is a UI convenience over the existing “additional read-only stores” feature, and in that case it should exist ~entirely in the top-level objects close to the UI. (I think this should be the case, but that’s not true after this PR.) And then the graph driver should be _completely unaware_ of the `options.ImageStore` option.

Or it is really a cross-cutting feature across the stores and the drivers, and in that case I think it should be _explicitly modeled_ as Go struct fields, and methods that can fail. The top-level store should, at the very least, _know_ whether the graph driver implements the “graphRoot additional image store” (maybe by the driver implementing a method that returns a value about that). And then, again, the store can just instruct the driver to use an additional image store, without the driver explicitly checking `options.ImageStore`.

It’s a strong “smell” of unclean design to me that `store` and the drivers (only one of them!), within the same codebase, communicate via `fmt.Sprintf("AdditionalImageStore=%s", options.ImageStore)`. Why use text when we we have Go struct fields?

File: store.go [hunk 5]
```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
...
+		imageStores = append(imageStores, s.rwImageStores...)
+		imageFound := false
+		primaryImageStore := true
+		for _, imageStore := range imageStores {
+			if !imageStore.Exists(id) {
+				primaryImageStore = false
+				continue
+			}
+			imageFound = true
+			if !primaryImageStore {
```
Comment (mtrmac) [inline store.go:2509]: Shouldn’t this just be `if imageStore != s.imageStore`?

- That’s _much_ much more explicit about what we need to check for. The locking logic _really_ doesn’t care about which store was used first, or whether the image exists.
- It’s much shorter to implement
- This one is actually incorrect: If an image exists in all stores, `primaryImageStore` is never set to `false`, and the other stores incorrectly be used without locking.

File: store.go [hunk 6]
```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
...
-			image, err := s.imageStore.Get(id)
+		// Perform delete image on primary imageStore but if image
+		// is not found on primary store then try delete on any additional
+		// write-able image store.
+		imageStores := []rwImageStore{s.imageStore}
+		imageStores = append(imageStores, s.rwImageStores...)
+		imageFound := false
+		primaryImageStore := true
+		for _, imageStore := range imageStores {
+			if !imageStore.Exists(id) {
```
Comment (mtrmac) [inline store.go:2504]: This is using `imageStore`, if it is not `s.imageStore`, completely without any locking; that’s incorrect.

File: store.go [hunk 7]
```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
 func (s *store) DeleteImage(id string, commit bool) (layers []string, err error) {
 	layersToRemove := []string{}
 	if err := s.writeToAllStores(func(rlstore rwLayerStore) error {
-		if s.imageStore.Exists(id) {
-			image, err := s.imageStore.Get(id)
+		// Perform delete image on primary imageStore but if image
+		// is not found on primary store then try delete on any additional
+		// write-able image store.
+		imageStores := []rwImageStore{s.imageStore}
+		imageStores = append(imageStores, s.rwImageStores...)
```
Comment (mtrmac) [inline store.go:2500]: (There’s enough precedent in this file for `for _, store := range append([]someStore{value}, otherValues…)`, without the extra variable.)

File: store.go [hunk 8]
```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
 func (s *store) DeleteImage(id string, commit bool) (layers []string, err error) {
 	layersToRemove := []string{}
 	if err := s.writeToAllStores(func(rlstore rwLayerStore) error {
-		if s.imageStore.Exists(id) {
-			image, err := s.imageStore.Get(id)
+		// Perform delete image on primary imageStore but if image
+		// is not found on primary store then try delete on any additional
+		// write-able image store.
```
Comment (mtrmac) [inline store.go:2498]: This comment would, typically, also be of the “code says it already” kind — except that this one is factually incorrect: the code, as is, deletes the image from _all_ image stores where it is found.

File: store.go [hunk 9]
```diff
@@ -2479,8 +2493,28 @@ func (s *store) DeleteLayer(id string) error {
...
+		imageStores := []rwImageStore{s.imageStore}
+		imageStores = append(imageStores, s.rwImageStores...)
+		imageFound := false
+		primaryImageStore := true
+		for _, imageStore := range imageStores {
```
Comment (mtrmac) [inline store.go:2503]: I would prefer just `store`, or maybe `is` or whatever, as the loop variable. It’s a _risk_ that `s.imageStore` and `imageStore` both exist, and mean different things.

**Comment by flouthoc** [comment]
I'll open a new PR to address these refactor comments.
